odoo/addons/base/models/ir_mail_server.py

```diff
@@ -511,6 +582,80 @@ def send_email(self, message, mail_server_id=None, smtp_server=None, smtp_port=N
             raise MailDeliveryException(_("Mail Delivery Failed"), msg)
         return message_id
 
+    @api.model
+    def _find_mail_server(self, email_from, mail_servers=None):
+        """Find the appropriate mail server for the given email address.
+
+        Returns: Record<ir.mail_server>, email_from
+        - Mail server to use to send the email (None if we use the odoo-bin arguments)
+        - Email FROM to use to send the email (in some case, it might be impossible
+          to use the given email address directly if no mail server is configured for)
+        """
+        email_from_normalized = email_normalize(email_from)
+        email_from_domain = email_domain_extract(email_from_normalized)
+        notifications_email = email_normalize(self._get_default_from_address())
+        notifications_domain = email_domain_extract(notifications_email)
+
+        if not mail_servers:
+            mail_servers = self.sudo().search([], order='sequence')
+
+        # 1. Try to find a mail server for the right mail from
+        mail_server = mail_servers.filtered(lambda m: email_normalize(m.from_filter) == email_from_normalized)
+        if mail_server:
+            return mail_server[0], email_from
```

Comment:
If `email_from` is `None` and there is a server with an unset `from_filter` is this not going to return that server and `None`?

Reply:
The email_from can not be None, because otherwise an error will be raised in `build_email` (so it's ok if find_mail_server return None for the FROM in that case)